tensorflow/python/ops/math_ops.py

```diff
@@ -4003,3 +4003,37 @@ def polyval(coeffs, x, name=None):
     for c in coeffs[1:]:
       p = c + p * x
     return p
+
+@tf_export("math.reciprocal_no_nan")
+def reciprocal_no_nan(x, name=None):
+  """Performs a safe reciprocal operation, element wise.
+  If a particular element is zero, the reciprocal for that element is
+  also set to zero.
+
+  For example:
+  ```python
+  x = tf.constant([2.0, 0.5, 0, 1], dtype=tf.float32)
+  tf.math.reciprocal_no_nan(x)  # [ 0.5, 2, 0.0, 1.0 ]
+  ```
+
+  Args:
+    x: A `Tensor` of type `float16`, `float32`, `float64`
+       `complex64` or `complex128`.
+    name: A name for the operation (optional).
+
+  Returns:
+    A `Tensor` of same shape and type as `x`.
+
+  Raises:
+    TypeError: x must be of a valid dtype.
+
+  """
+  allowed_dtypes = [dtypes.float16, dtypes.float32, dtypes.float64,
+                    dtypes.complex64, dtypes.complex128]
+  with ops.name_scope(name, "reciprocal_no_nan", [x]) as scope:
+    x = ops.convert_to_tensor(x, name="x")
+    if x.dtype.base_dtype not in allowed_dtypes:
+      raise TypeError("x has incorrect data type: {} \n "
```

Comment:
Does this error message add value over just using the standard error message you'll get out of the wrong dtype for div_no_nan?

If you do insist on keeping this error message it should say "unsupported" instead of incorrect, and probably mention reciprocal_no_nan.

Reply:
That was going to be my other PR/issue: currently `div_no_nan()` (and for that matter `reciprocal()`) does not return the `TypeError` or `ValueError` exceptions when using eager execution. For that reason I added this in this function. 
When I currently run `tf.math.divide_no_nan()` on eager mode with an invalid datatype (int32), I get the following exception:
```
NotFoundError: Could not find valid device for node.
Node:{{node DivNoNan}}
All kernels registered for op DivNoNan :
  device='XLA_CPU'; T in [DT_FLOAT, DT_DOUBLE, DT_COMPLEX64, DT_COMPLEX128, DT_HALF]
  device='XLA_CPU_JIT'; T in [DT_FLOAT, DT_DOUBLE, DT_COMPLEX64, DT_COMPLEX128, DT_HALF]
  device='CPU'; T in [DT_HALF]
  device='CPU'; T in [DT_FLOAT]
  device='CPU'; T in [DT_DOUBLE]
  device='CPU'; T in [DT_COMPLEX64]
  device='CPU'; T in [DT_COMPLEX128]
 [Op:DivNoNan] name: div_no_nan/
```
It seems to work fine in non-eager mode. 

I can make the change you suggested ~incorrect~->unsupported in the meanwhile.

Reply:
Most of the unit tests I ran were in eager mode for convenience and so I added this bit of exception handling. I agree however since this function just uses `divide_no_nan()` we should catch the same exception and pass it over, avoid duplicate code.

Reply:
@alextp can you suggest if my approach is fine?

Reply:
exactly, let's let divide_no_nan raise the exceptions

Reply:
@alextp so I'll remove exception handling here. But in eager mode, div_no_nan() doesn't give a valid/useful exception. How should we handle that? I can also raise a separate issue to discuss this.